modules/core/src/main/java/org/apache/ignite/internal/processors/service/ServicesDeploymentManager.java

```diff
@@ -0,0 +1,553 @@
+     * client reconnect events.
+     *
+     * @param evt Discovery event.
+     * @param discoCache Discovery cache.
+     * @param depActions Service deployment actions.
+     */
+    protected void onLocalJoin(DiscoveryEvent evt, DiscoCache discoCache, ServicesDeploymentActions depActions) {
+        checkClusterStateAndAddTask(evt, discoCache, depActions);
+    }
+
+    /**
+     * Invokes {@link GridWorker#blockingSectionBegin()} for service deployment worker.
+     * <p/>
+     * Should be called from service deployment worker thread.
+     */
+    protected void deployerBlockingSectionBegin() {
+        assert depWorker != null && Thread.currentThread() == depWorker.runner();
+
+        depWorker.blockingSectionBegin();
+    }
+
+    /**
+     * Invokes {@link GridWorker#blockingSectionEnd()} for service deployment worker.
+     * <p/>
+     * Should be called from service deployment worker thread.
+     */
+    protected void deployerBlockingSectionEnd() {
+        assert depWorker != null && Thread.currentThread() == depWorker.runner();
+
+        depWorker.blockingSectionEnd();
+    }
+
+    /**
+     * Checks cluster state and handles given event.
+     * <pre>
+     * - if cluster is active, then adds event in deployment queue;
+     * - if cluster state in transition, them adds to pending events;
+     * - if cluster is inactive, then ignore event;
+     * </pre>
+     * <b>Should be called from discovery thread.</b>
+     *
+     * @param evt Discovery event.
+     * @param discoCache Discovery cache.
+     * @param depActions Services deployment actions.
+     */
+    private void checkClusterStateAndAddTask(@NotNull DiscoveryEvent evt, @NotNull DiscoCache discoCache,
+        @Nullable ServicesDeploymentActions depActions) {
+        if (discoCache.state().transition())
+            pendingEvts.add(new PendingEventHolder(evt, discoCache.version(), depActions));
+        else if (discoCache.state().active())
+            addTask(evt, discoCache.version(), depActions);
+        else if (log.isDebugEnabled())
+            log.debug("Ignore event, cluster is inactive, evt=" + evt);
+    }
+
+    /**
+     * Adds deployment task with given deployment process id.
+     * </p>
+     * <b>Should be called from discovery thread.</b>
+     *
+     * @param evt Discovery event.
+     * @param topVer Topology version.
+     * @param depActions Services deployment actions.
+     */
+    private void addTask(@NotNull DiscoveryEvent evt, @NotNull AffinityTopologyVersion topVer,
+        @Nullable ServicesDeploymentActions depActions) {
+        final ServicesDeploymentProcessId depId = deploymentId(evt, topVer);
+
+        ServicesDeploymentTask task = tasks.computeIfAbsent(depId,
+            t -> new ServicesDeploymentTask(ctx, depId));
+
+        if (!task.onAdditionInQueue()) {
+            log.warning("Do not start service deployment process for event: " + evt);
```

Comment:
This is not informative. What is the reason? What should a user do?

Reply:
This is not useful for the user. I changed the level to 'debug' and enrol description.